pkg/kubernetes/deployment/appsv1/v1alpha1/rollout_status.go

```diff
@@ -0,0 +1,152 @@
+// each predicate checks. It uses predicateName as key.
+var rolloutStatuses = map[predicateName]rolloutStatus{
+	// PredicateProgressDeadlineExceeded refer to rolloutStatus
+	// for predicate IsProgressDeadlineExceeded.
+	PredicateProgressDeadlineExceeded: func(d *deploy) string {
+		return "Deployment exceeded its progress deadline"
+	},
+	// PredicateOlderReplicaActive refer to rolloutStatus for
+	// predicate IsOlderReplicaActive.
+	PredicateOlderReplicaActive: func(d *deploy) string {
+		if d.object.Spec.Replicas == nil {
+			return "Replica update in progress : some older replicas have been updated"
+		}
+		return fmt.Sprintf(
+			"Replica update in progress : %d out of %d new replicas have been updated",
+			d.object.Status.UpdatedReplicas, *d.object.Spec.Replicas)
+	},
+	// PredicateTerminationInProgress refer rolloutStatus
+	// for predicate IsTerminationInProgress.
+	PredicateTerminationInProgress: func(d *deploy) string {
+		return fmt.Sprintf(
+			"Replica termination in progress : %d old replicas are pending termination",
+			d.object.Status.Replicas-d.object.Status.UpdatedReplicas)
+	},
+	// PredicateUpdateInProgress refer to rolloutStatus for predicate IsUpdateInProgress.
+	PredicateUpdateInProgress: func(d *deploy) string {
+		return fmt.Sprintf(
+			"Replica update in progress : %d of %d updated replicas are available",
+			d.object.Status.AvailableReplicas, d.object.Status.UpdatedReplicas)
+	},
+	// PredicateNotSpecSynced refer to status rolloutStatus for predicate IsNotSyncSpec.
+	PredicateNotSpecSynced: func(d *deploy) string {
+		return "Deployment rollout in-progress : deployment rollout in-progress : waiting for deployment spec update to be observed"
+	},
+}
+
+// rolloutChecks contains a group of predicate it uses predicateName as key.
+var rolloutChecks = map[predicateName]predicate{
+	// PredicateProgressDeadlineExceeded refer to predicate IsProgressDeadlineExceeded.
+	PredicateProgressDeadlineExceeded: IsProgressDeadlineExceeded(),
+	// PredicateOlderReplicaActive refer to predicate IsOlderReplicaActive.
+	PredicateOlderReplicaActive: IsOlderReplicaActive(),
+	// PredicateTerminationInProgress refer to predicate IsTerminationInProgress.
+	PredicateTerminationInProgress: IsTerminationInProgress(),
+	// PredicateUpdateInProgress refer to predicate IsUpdateInProgress.
+	PredicateUpdateInProgress: IsUpdateInProgress(),
+	// PredicateNotSpecSynced refer to predicate IsSyncSpec.
+	PredicateNotSpecSynced: IsNotSyncSpec(),
+}
+
+// rolloutOutput struct contains message and boolean value to show rolloutstatus
+type rolloutOutput struct {
+	IsRolledout bool   `json:"isRolledout"`
+	Message     string `json:"message"`
+}
+
+// rawFn is a typed function that abstracts
+// conversion of rolloutOutput struct to raw byte
+type rawFn func(r *rolloutOutput) ([]byte, error)
+
+// rawFn is a typed function that abstracts
+// conversion of rolloutOutput struct
+type asRolloutOutputFn func(r *rolloutOutput) (*rolloutOutput, error)
```

Comment:
IMO we can reuse `rolloutBuildOption` here as well.
The pointer that is provided as input can be used by the caller.
error can be used as a value if required.
This error can be returned along with the final function e.g.

```go
func (r *rollout) Raw() ([]bytes, error) {
  if r.isError() {
    return nil, errors.Errorf("%v", r.errs)
  }
  // ...
}


type rollout struct {
    // ...
    errs []error
}

func (r *rollout) isError() bool {
  return len(r.errs) > 0 
}

func (r *rollout) Errors() []error {
  return r.errs
}
```